### PR TITLE
Authentication implementation.

### PR DESCRIPTION
**Pull request by sadcenter**
# Description

Changes auth implementation from yggdrasil one to mine one. (toggleable in configuration)
It uses https://github.com/Electroid/mojang-api and caches the player's texture to the file.

Fixes/Adds #127 (Not completely due it's impossible, more info in `Additional comments` section.)

# How has this been tested?

Tested by generating skulls and toggling `online-mode`

# Additional Comments

It's impossible to remove the delay completely in async action (sync would be even worse). However, we can try to reduce the delay by using better API, cache, etc. 
Default cache time is 1 day.
If there's something wrong with changes, let me know.

# Checklist:

- [X] I have reviewed my code thoroughly.
- [X] I have tested my code. (I recommend testing it again, it's a big update so I might have skipped something)
- [X] My changes generate no new warnings

### CODE REVIEW

File: NachoSpigot-Server/src/main/java/me/elier/nachospigot/config/NachoConfig.java [hunk 1]
```diff
@@ -141,6 +141,7 @@ static void loadComments() {
         c.addComment("settings.commands.permissions.version", "Enables a required permission to use /version");
         c.addComment("settings.commands.permissions.plugins", "Enables a required permission to use /plugins");
         c.addComment("settings.commands.enable-help-command", "Toggles the /help command");
+        c.addComment("settings.use-nacho-authenticator", "Enables nacho authentication");
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/me/elier/nachospigot/config/NachoConfig.java:144]:
```suggestion
        c.addComment("settings.use-nacho-authenticator", "Enables our own authentication system which uses Electroid's Mojang API.");
```

File: NachoSpigot-Server/src/main/java/net/minecraft/server/DedicatedServer.java [hunk 2]
```diff
@@ -178,7 +181,12 @@ public void run() {
             // Spigot start
             NachoConfig.init((File) options.valueOf("nacho-settings")); // NachoSpigot - Load config before PlayerList
             KnockbackConfig.init((File) options.valueOf("knockback-settings"));
-            this.setPlayerList(new DedicatedPlayerList(this)); // Nacho - deobfuscate setPlayerList
+            YggdrasilAuthenticationService yggdrasilAuthenticationService = new YggdrasilAuthenticationService(super.e, UUID.randomUUID().toString());
+            this.V = NachoConfig.useNachoAuthenticator ? new NachoAuthenticatorService(yggdrasilAuthenticationService) : yggdrasilAuthenticationService;
+            this.W = this.V.createMinecraftSessionService();
+            this.Y = this.V.createProfileRepository();
+
+            this.setPlayerList(new DedicatedPlayerList(this));
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/net/minecraft/server/DedicatedServer.java:189]:
```suggestion
            // Nacho start - Use our own authentication system
            YggdrasilAuthenticationService yggdrasilAuthenticationService = new YggdrasilAuthenticationService(super.e, UUID.randomUUID().toString());
            this.V = NachoConfig.useNachoAuthenticator ? new NachoAuthenticatorService(yggdrasilAuthenticationService) : yggdrasilAuthenticationService;
            this.W = this.V.createMinecraftSessionService();
            this.Y = this.V.createProfileRepository();
            // Nacho end

            this.setPlayerList(new DedicatedPlayerList(this));
```

File: NachoSpigot-Server/src/main/java/net/minecraft/server/ItemSkull.java [hunk 3]
```diff
@@ -121,14 +126,19 @@ public boolean a(final NBTTagCompound nbttagcompound) { // Spigot - make final
...
+                });
+            } else {
+                TileEntitySkull.b(gameprofile, new Predicate<GameProfile>() {
+                    @Override
+                    public boolean apply(@Nullable GameProfile gameProfile) {
+                        nbttagcompound.set("SkullOwner", GameProfileSerializer.serialize(new NBTTagCompound(), gameProfile));
+                        return false;
+                    }
+                });
+            }
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/net/minecraft/server/ItemSkull.java:141]:
```suggestion
            // Nacho start - Use our own authentication system
            if(NachoConfig.useNachoAuthenticator) {
                ((NachoAuthenticatorService) MinecraftServer.getServer().getAuthenticator()).getProfile(gameprofile.getName()).thenAccept(profile -> {
                    nbttagcompound.set("SkullOwner", GameProfileSerializer.serialize(new NBTTagCompound(), profile));
                });
            } else {
                TileEntitySkull.b(gameprofile, new Predicate<GameProfile>() {
                    @Override
                    public boolean apply(@Nullable GameProfile gameProfile) {
                        nbttagcompound.set("SkullOwner", GameProfileSerializer.serialize(new NBTTagCompound(), gameProfile));
                        return false;
                    }
                });
            }
            // Nacho end
```

File: NachoSpigot-Server/src/main/java/net/minecraft/server/MinecraftServer.java [hunk 4]
```diff
@@ -799,6 +797,9 @@ public void B() {
             processQueue.remove().run();
         }
         SpigotTimings.processQueueTimer.stopTiming(); // Spigot
+        if(NachoConfig.useNachoAuthenticator) {
+            ((NachoAuthenticatorService) this.V).getProfileCache().setTicked(false);
+        }
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/net/minecraft/server/MinecraftServer.java:802]:
```suggestion
        // Nacho start
        if(NachoConfig.useNachoAuthenticator) {
            ((NachoAuthenticatorService) this.V).getProfileCache().setTicked(false);
        }
        // Nacho end
```

File: NachoSpigot-Server/src/main/java/net/minecraft/server/MinecraftServer.java [hunk 5]
```diff
@@ -1424,6 +1425,10 @@ public int getMaxBuildHeight() {
         return this.F;
     }
 
+    public AuthenticationService getAuthenticator() {
+        return this.V;
+    }
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/net/minecraft/server/MinecraftServer.java:1430]:
```suggestion
    // Nacho start
    public AuthenticationService getAuthenticator() {
        return this.V;
    }
    // Nacho end
```

File: NachoSpigot-Server/src/main/java/net/minecraft/server/TileEntitySkull.java [hunk 6]
```diff
@@ -137,23 +135,30 @@ public void setGameProfile(GameProfile gameprofile) {
     }
 
     private void e() {
-        // Spigot start
+        // Nacho start
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/net/minecraft/server/TileEntitySkull.java:138]:
```suggestion
        // Spigot start
```

File: NachoSpigot-Server/src/main/java/net/minecraft/server/TileEntitySkull.java [hunk 7]
```diff
@@ -137,23 +135,30 @@ public void setGameProfile(GameProfile gameprofile) {
...
         GameProfile profile = this.g;
         setSkullType( 0 ); // Work around client bug
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/net/minecraft/server/TileEntitySkull.java:140]:
```suggestion
        setSkullType( 0 ); // Work around client bug
        // Nacho start - Use our own authentication system
```

File: NachoSpigot-Server/src/main/java/net/minecraft/server/TileEntitySkull.java [hunk 8]
```diff
@@ -137,23 +135,30 @@ public void setGameProfile(GameProfile gameprofile) {
...
+            g = input;
+            update();
+            if (world != null) {
+                world.notify(position);
             }
-        }); 
-        // Spigot end
+            return false;
+        });
+        // Nacho end
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/net/minecraft/server/TileEntitySkull.java:161]:
```suggestion
        // Spigot/Nacho end
```

File: NachoSpigot-Server/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java [hunk 9]
```diff
@@ -73,22 +70,27 @@ void applyToItem(final NBTTagCompound tag) { // Spigot - make final
...
+            setSkullNbt(tag, profile);
+
+            if (NachoConfig.useNachoAuthenticator) {
+                NachoAuthenticatorService authenticator = (NachoAuthenticatorService) MinecraftServer.getServer().getAuthenticator();
+
+                authenticator.getProfile(profile.getName()).thenAccept(gameProfile -> {
+                    setSkullNbt(tag, gameProfile);
+                });
+            } else TileEntitySkull.b(profile, gameProfile -> {
+                setSkullNbt(tag, gameProfile);
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java:82]:
```suggestion
            // Nacho start - Use our own authentication system
            setSkullNbt(tag, profile);

            if (NachoConfig.useNachoAuthenticator) {
                NachoAuthenticatorService authenticator = (NachoAuthenticatorService) MinecraftServer.getServer().getAuthenticator();

                authenticator.getProfile(profile.getName()).thenAccept(gameProfile -> {
                    setSkullNbt(tag, gameProfile);
                });
            } else TileEntitySkull.b(profile, gameProfile -> { // Spigot - do an async lookup of the profile
                setSkullNbt(tag, gameProfile);
                // Nacho end
```

File: NachoSpigot-Server/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java [hunk 10]
```diff
@@ -73,22 +70,27 @@ void applyToItem(final NBTTagCompound tag) { // Spigot - make final
...
             });
-            // Spigot end
         }
     }
 
+    private void setSkullNbt(NBTTagCompound tag, GameProfile gameProfile) {
+        NBTTagCompound ownerTag = new NBTTagCompound();
+        GameProfileSerializer.serialize(ownerTag, gameProfile);
+        tag.set(SKULL_OWNER.NBT, ownerTag);
+    }
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java:92]:
```suggestion
    // Nacho start
    private void setSkullNbt(NBTTagCompound tag, GameProfile gameProfile) {
        NBTTagCompound ownerTag = new NBTTagCompound();
        GameProfileSerializer.serialize(ownerTag, gameProfile);
        tag.set(SKULL_OWNER.NBT, ownerTag);
    }
    // Nacho end
```

File: NachoSpigot-Server/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java [hunk 11]
```diff
@@ -126,18 +128,20 @@ public boolean setOwner(String name) {
...
+        EntityPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(name);
+        if (profile == null && player != null) profile = player.getProfile();
+        // PaperSpigot end
+
         if (profile == null) {
         	// name.toLowerCase(java.util.Locale.ROOT) causes the NPE
-        	profile = TileEntitySkull.skinCache.getIfPresent(name.toLowerCase(java.util.Locale.ROOT)); // Paper // tries to get from skincache
+        	profile = NachoConfig.useNachoAuthenticator ?
+                    ((NachoAuthenticatorService) MinecraftServer.getServer().getAuthenticator()).getPresentProfile(name) :
+                    TileEntitySkull.skinCache.getIfPresent(name);
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java:141]:
```suggestion
                // Nacho start - Use our own authentication system
        	profile = NachoConfig.useNachoAuthenticator ?
                    ((NachoAuthenticatorService) MinecraftServer.getServer().getAuthenticator()).getPresentProfile(name.toLowerCase(java.util.Locale.ROOT)) :
                    TileEntitySkull.skinCache.getIfPresent(name.toLowerCase(java.util.Locale.ROOT)); // Paper // tries to get from skincache
                // Nacho end
```

**Review by ghost — COMMENTED** [review]
Looks good! This is just a brief review, I'll do a full review and test when I can. Also you should try to not edit the formatting of the files, I see that there's a lot of files with just formatting changes.

**Review by ghost — APPROVED** [review]

**Comment by Blyrex** [comment]
its not a good idea to use private apis in this project

File: NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/NachoAuthenticatorService.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class NachoAuthenticatorService implements AuthenticationService {
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/NachoAuthenticatorService.java:39]: This class should be called `NachoAuthenticationService`

File: NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/storage/ProfileCache.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        this.save();
+    }
+
+    private Map<String, CachedProfile> load() {
+        if (!CACHE_FILE.exists()) {
+            try {
+                CACHE_FILE.createNewFile();
+                FileUtils.writeStringToFile(CACHE_FILE, "{}");
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/storage/ProfileCache.java:39]:
```suggestion
                Bukkit.getLogger().log(Level.SERVERE, "Error while creating texturecache.json", e);
```

File: NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/storage/ProfileCache.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import net.minecraft.server.MinecraftServer;
+import org.apache.commons.io.FileUtils;
+import org.spigotmc.CaseInsensitiveMap;
+
+import java.io.*;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/storage/ProfileCache.java:14]:
```suggestion
import java.util.concurrent.Executors;
import java.util.logging.Level;
```

File: NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/storage/ProfileCache.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
+package com.github.sadcenter.auth.storage;
+
+import com.github.sadcenter.auth.NachoAuthenticatorService;
+import com.google.common.reflect.TypeToken;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import net.minecraft.server.MinecraftServer;
+import org.apache.commons.io.FileUtils;
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/storage/ProfileCache.java:7]:
```suggestion
import org.apache.commons.io.FileUtils;
import org.bukkit.Bukkit;
```

File: NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/storage/ProfileCache.java [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+                FileUtils.writeStringToFile(CACHE_FILE, "{}");
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        } else try (BufferedReader fileReader = new BufferedReader(new FileReader(CACHE_FILE))) {
+            Map<String, CachedProfile> loadedCache = NachoAuthenticatorService.GSON
+                    .fromJson(fileReader, new TypeToken<HashMap<String, CachedProfile>>() {}.getType());
+            return this.filter(loadedCache);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/storage/ProfileCache.java:46]:
```suggestion
            Bukkit.getLogger().log(Level.SERVERE, "Error while loading profile caches", e);
```

File: NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/session/NachoSessionService.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+
+import com.github.sadcenter.auth.NachoAuthenticatorService;
+import com.github.sadcenter.auth.response.HasJoinedServerResponse;
+import com.github.sadcenter.auth.utils.HashMapBuilder;
+import com.google.common.collect.Iterables;
+import com.mojang.authlib.GameProfile;
+import com.mojang.authlib.minecraft.MinecraftProfileTexture;
+import com.mojang.authlib.minecraft.MinecraftSessionService;
+import com.mojang.authlib.properties.Property;
+import com.mojang.authlib.properties.PropertyMap;
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/session/NachoSessionService.java:11]:
```suggestion
import com.mojang.authlib.properties.PropertyMap;
import org.bukkit.Bukkit;
```

File: NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/session/NachoSessionService.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    public void joinServer(GameProfile gameProfile, String authToken, String serverId) {
+        Map<String, Object> request = new HashMapBuilder()
+                .withParam("accessToken", authToken)
+                .withParam("selectedProfile", gameProfile.getId())
+                .withParam("serverId", serverId).getMap();
+
+        try {
+            this.authenticator.fetchPost(NachoAuthenticatorService.url(JOIN_URL), NachoAuthenticatorService.json(request));
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/session/NachoSessionService.java:38]:
```suggestion
            Bukkit.getLogger().log(Level.SERVERE, "Error loading profile " + gameProfile.getName() + " (UUID: " + gameProfile.getId() + ")", e);
```

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Comment by sadcenter** [comment]
> its not a good idea to use private apis in this project

I personally trust this API. I haven't ever met with a situation when this API is down (I'm using this in every project that needs data from Mojang). But not everyone has to trust it, that's why I added an option in the configuration to disable that.
IMO it would be great to add an option that allows you to toggle between ashcon API and Mojang's API

**Comment by Ytnoos** [comment]
> > its not a good idea to use private apis in this project
> 
> 
> 
> I personally trust this API. I haven't ever met with a situation when this API is down (I'm using this in every project that needs data from Mojang). But not everyone has to trust it, that's why I added an option in the configuration to disable that.
> 
> IMO it would be great to add an option that allows you to toggle between ashcon API and Mojang's API

I think it would be good to automatically switch to Mojang's API if any exception is caught with ashcon API

**Review by ghost — APPROVED** [review]

**Comment by CyberFlameGO** [comment]
> its not a good idea to use private apis in this project

The API seems open source (as declared on the issue init message) and claims to be hosted on Cloudflare Workers

**Comment by CyberFlameGO** [comment]
Oh also the person running and maintaining the API works at Cloudflare

File: NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/NachoAuthenticationService.java [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+        return null;
+    }
+
+    public GameProfile getPresentProfile(String name) {
+        CompletableFuture<GameProfile> profile = this.gameProfileCache.getIfPresent(name);
+        return profile == null ? null :
+                profile.getNow(null);
+    }
+
+    public CompletableFuture<UUID> getUuid(String name) {
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/com/github/sadcenter/auth/NachoAuthenticationService.java:121]: Should be `getUuidFromApi`

File: NachoSpigot-Server/src/main/java/me/elier/nachospigot/config/NachoConfig.java [hunk 20]
```diff
@@ -390,4 +392,16 @@ private static void enableProtocolShim() {
...
+
+    private static void useNachoAuthenticator() {
+        useNachoAuthenticator = getBoolean("settings.authenticator.use-nacho-authenticator", false);
+    }
+
+    public static boolean alwaysUseMojang;
+
+    private static void alwaysUseMojang() {
+        alwaysUseMojang = getBoolean("settings.authenticator.always-use-mojang", true);
+    }
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/me/elier/nachospigot/config/NachoConfig.java:406]:
```suggestion
    public static boolean useNachoAuthenticator;
    public static boolean alwaysUseMojang;

    private static void authentication() {
        useNachoAuthenticator = getBoolean("settings.authentication.use-nacho-authenticator", false);
        alwaysUseMojang = getBoolean("settings.authentication.always-use-mojang", true);
    }
```

File: NachoSpigot-Server/src/main/java/me/elier/nachospigot/config/NachoConfig.java [hunk 21]
```diff
@@ -141,6 +141,8 @@ static void loadComments() {
         c.addComment("settings.commands.permissions.version", "Enables a required permission to use /version");
         c.addComment("settings.commands.permissions.plugins", "Enables a required permission to use /plugins");
         c.addComment("settings.commands.enable-help-command", "Toggles the /help command");
+        c.addComment("settings.authenticator.use-nacho-authenticator", "Enables our own authentication system which uses Electroid's Mojang API.");
+        c.addComment("settings.authenticator.always-use-mojang", "Force to use mojang api no matter what");
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/me/elier/nachospigot/config/NachoConfig.java:145]:
```suggestion
        c.addComment("settings.authenticator.use-nacho-authentication", "Enables our own authentication system.");
        c.addComment("settings.authenticator.always-use-mojang", "Force to use mojang api no matter what");
```

File: NachoSpigot-Server/src/main/java/me/elier/nachospigot/config/NachoConfig.java [hunk 22]
```diff
@@ -390,4 +392,16 @@ private static void enableProtocolShim() {
...
+    public static boolean useNachoAuthenticator;
+
+    private static void useNachoAuthenticator() {
+        useNachoAuthenticator = getBoolean("settings.authenticator.use-nacho-authenticator", false);
+    }
+
+    public static boolean alwaysUseMojang;
+
+    private static void alwaysUseMojang() {
+        alwaysUseMojang = getBoolean("settings.authenticator.always-use-mojang", true);
```
Comment (Elierrr) [inline NachoSpigot-Server/src/main/java/me/elier/nachospigot/config/NachoConfig.java:405]: Imo this should be inverted and be use-ashcon-api

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by Lucaskyy** [comment]
Looks good! Awesome PR :)
Since some reviews are not yet resolved, I'll wait until this is fully ready. Since I do not look at this repo often because I don't have much time, please ping me if possible so I can merge it :)

**Comment by Lucaskyy** [comment]
Also, since I merged some PRs please fix the conflicts if possible :)

**Comment by sadcenter** [comment]
oops Intelij IDEA kinda messed up I believe

**Comment by Elierrr** [comment]
> oops Intelij IDEA kinda messed up I believe

Rebasing on master should fix it

**Comment by Lucaskyy** [comment]
Is this PR stale or not?

**Comment by HowardZHY** [comment]
is [https://github.com/CobbleSword/NachoSpigot/issues/353](https://github.com/CobbleSword/NachoSpigot/issues/353) related to this?

**Comment by sadcenter** [comment]
> is #353 related to this?
it **might** fix it, but it's unlikely possible since authlib wasn't touched in nacho

**Comment by sadcenter** [comment]
> Is this PR stale or not?

yea, I will resolve the conflicts tomorrow (if tram won't hit me or smth)

**Comment by sadcenter** [comment]
CodeQL uses Java 11 where `exceptionallyCompose` and `exceptionallyComposeAsync` doesn't exist :/

**Pull request closed without merging**